L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,504 @@
+  absl::string_view peer_cert_full_chain() const;
+
+  // The common name string from the Subject extension in the peer's leaf
+  // certificate.
+  absl::string_view common_name() const;
+
+  // The subject name of the root certificate used to verify the peer chain
+  // If verification fails or the peer certificate is self-signed, this will be an
+  // empty string. If verification is successful, it is a comma-separated list,
+  // where the entries are of the form "FIELD_ABBREVIATION=string"
+  // ex: "CN=testca,O=Internet Widgits Pty Ltd,ST=Some-State,C=AU"
+  // ex: "CN=GTS Root R1,O=Google Trust Services LLC,C=US"
+  absl::string_view verified_root_cert_subject() const;
+
+  // The (possibly empty) list of URI names from the Subject Alternative Name
+  // extension in the peer's leaf certificate.
+  std::vector<absl::string_view> uri_names() const;
+
+  // The (possibly empty) list of DNS names from the Subject Alternative Name
+  // extension in the peer's leaf certificate.
+  std::vector<absl::string_view> dns_names() const;
+
+  // The (possibly empty) list of email names from the Subject Alternative Name
+  // extension in the peer's leaf certificate.
+  std::vector<absl::string_view> email_names() const;
+
+  // The (possibly empty) list of IP names from the Subject Alternative Name
+  // extension in the peer's leaf certificate.
+  std::vector<absl::string_view> ip_names() const;
+}
+
+// The base class of all verifier implementations. Note that custom
+// verifier implementations can compose their functionality with existing
+// implementations of this interface, such as HostNameVerifier, by delegating
+// to an instance of that class.
+class CertificateVerifierInterface {
+ public:
+  virtual ~CertificateVerifierInterface() = default;
+
+  // Performs a custom verification check.
+  // Returns true if verification occurs synchronously, in which case the
+  // verification result is populated in the sync_status output parameter.
+  // Returns false if verification occurs asynchronously, in which case the
+  // callback must be called when verification is complete and the verification
+  // result must be populated in the sync_status output parameter.
+  //
+  // request: the verification information associated with this request
+  // callback: This will only take effect if the verifier is asynchronous.
+  //           The function that gRPC will invoke when the verifier has already
+  //           completed its asynchronous check. Callers can use this function
+  //           to perform any additional checks. The input parameter of the
+  //           std::function indicates the status of the verifier check.
+  // sync_status: This will only be useful if the verifier is synchronous.
+  //              The status of the verifier as it has already done it's
+  //              synchronous check.
+  // return: return true if executed synchronously, otherwise return false
+  virtual bool Verify(TlsCustomVerificationCheckRequest* request,
```

Comment:
From review, change this to return an absl::StatusOr<bool> instead of happy the Status output parameter?

Reply:
I think we should look at this some more, the bool return is `true` if sync, `false` otherwise. I'm not sure what we gain with a `StatusOr`?

Reply:
I think what we gain is being able to remove the `absl::Status* sync_status` output parameter, no?

Reply:
For example, if the signature becomes `absl::StatusOr<bool> Verify(request, cb)`, then the semantics are:
- If error status, failed synchronously.
- If OK status and true, succeed synchronously.
- If OK status and false, verification still in progress asynchronously.

This is nice because you only have to check the value of one object (the output variable) to understand what is happening, rather than two objects (the output variable and the output parameter). This also avoids easy ways for users to shoot themselves in the foot, c.f. forgetting to set sync_status.

Reply:
Ah that makes sense, changed, PTAL